### PR TITLE
[aot_compile] Add torch.compiler.precompile

### PR DESCRIPTION
**Pull request by bobrenjc93**
## Summary

Adds `torch.compiler.precompile`, an ahead-of-time precompile API that returns python code + serializable kernel cache that can be used to bypass kernel JIT compilation.

```python
# fn is the whole computation -- here a training step: forward, loss, backward.
train_step = lambda model, x, target: loss_fn(model(x), target).backward()

python_code, cache = torch.compiler.precompile(train_step, model, x, target)   # str, bytes
f_c = torch.compiler.precompile.load(python_code, cache)
f_c(model, x, target)   # pass the model again; grads accumulate on model.parameters()
```

## Backend contract

`torch.compiler.precompile` lowers through a two-layer "graph in -> (python, cache) out" contract:

- **`torch._inductor.compile_to_python(gm, inputs) -> (python, cache)`** -- the inner Inductor output (kernels + `call`) for the post-AOTAutograd dense graph. No prelude/epilogue.
- **`torch._functorch.aot_autograd.compile_to_python(gm, inputs) -> (python, cache)`** -- composes AOTAutograd's **own codegen'd runtime wrappers** (the prelude/epilogue) around the inner `call` into one self-contained module, rather than reimplementing them:
  - a capture hook (`capture_generated_sources` in `_aot_autograd/subclass_codegen.py`) retains each wrapper's source during compile;
  - the composer re-execs each wrapper with its resolved globals (inner call / sibling wrapper / public helper / baked metadata reconstructed as plain source), mirroring the live runtime (no name collisions, no source rewriting);
  - baked metadata (tensor-subclass placements/spec, view-replay recipes, ...) is emitted as plain Python source rather than embedded as a pickle blob, so the generated module contains no `pickle.loads` and is fully auditable; a value that cannot be expressed as source raises rather than falling back to an opaque embedding.

## Tracer

How `fn` is captured is selected by `tracer=`, a separate axis from the backend:
- `tracer="make_fx"` (default, and the only tracer implemented today) -- a NON-STRICT make_fx trace: it records the ATen ops that actually run when `fn` executes once on the example inputs and does not analyze your Python, so control flow and shapes are specialized to the example inputs (see the programming model). This is the source of the explicit contract callers must uphold.
- `tracer="dynamo"` (planned) -- in the future we plan to support a Dynamo-based tracer that analyzes the Python (graph breaks, guards) instead of specializing to a single traced path; it raises `NotImplementedError` today.

## Backends

`torch.compiler.precompile(fn, *args, backend=...)`:
- `backend="inductor"` (default) -- lowers through `aot_autograd.compile_to_python` (above).
- `backend="eager"` -- captures and runs the ATen graph as-is without Inductor (analogous to `torch.compile(backend="eager")`), for inspecting/debugging exactly what was traced. `python_code` inlines the readable, executable graph; the cache is empty (the eager backend has no kernels to accelerate, so `python_code` is the whole artifact).

## Self-contained artifact; cache is only an acceleration

`python_code` runs on its own (JIT) and is the single source of truth for the calling convention; `load` reads that back by parsing the source (an AST scrape, not an exec). The `cache` holds only the compiled/captured artifact and only lets `load` skip JIT (FxGraphCache hit) -- same results either way. The calling-convention metadata is single-sourced in `python_code`, and the cache envelope carries a format/version/backend tag plus a `code_hash` (sha256 of the exact `python_code` it accelerates) that `load` verifies: a stale-format, wrong-backend, or MISMATCHED (`python_code`, `cache`) pairing is rejected with a clear `PrecompileError` rather than silently running the cache's graph under the wrong code.

## Runtime safety checks

Because a precompiled artifact cannot recompile, every runtime call validates its inputs against what was traced and fails LOUDLY (`PrecompileError`) rather than silently mis-computing: a runtime input's shape, dtype, device, and (inductor backend) memory format must match the example; a `mark_unbacked` dim's declared `min`/`max` are enforced; and the runtime model's parameter/buffer structure -- names, order, each param/buffer's shape and dtype, AND (inductor backend) each weight's memory format -- must match the traced model (so a reordered, structurally-different, or same-name-but-wrong-shape model is rejected rather than scattering grads onto the wrong slot or silently mis-computing). `backend="eager"` is layout-flexible (no `assert_size_stride`) but still checks input shape/dtype/device and the full param/buffer structure. Both `python_code` and `cache` are untrusted EXECUTABLE input to `load` (it execs/unpickles), which warns on every load. The error type is `torch.compiler.PrecompileError`.

## Example: generated `python_code` showing the composed runtime wrappers

Two contrived examples (intentionally not realistic models, on CUDA) whose generated artifacts together exercise every runtime wrapper precompile composes -- tensor-subclass wrap/unwrap, functionalized RNG, output-alias regeneration, input-mutation reflection, and the orchestration -- plus the training grad-harvest path. (A tensor subclass and functionalized RNG cannot share one graph -- AOTAutograd asserts on the rng-state args -- so they are split across the two examples.) Each generated artifact is self-contained, with three sections: (1) inlined Inductor kernels + the composed AOTAutograd wrappers exposing `call`, (2) calling-convention metadata (`PARAM_NAMES`, `OUT_SPEC`, ...), and (3) a `forward(*args)` driver that lifts the runtime model's params/buffers (no weights baked) and scatters any harvested grads onto `model.parameters().grad`.

Each case links the **generator** (builds the model + calls `torch.compiler.precompile`) and the resulting **generated `python_code`** (pastry links for now):

- **Inference (tensor subclass + input mutation + output aliasing)**: [generator](https://www.internalfb.com/intern/paste/P2389160270/) -> [generated `python_code`](https://www.internalfb.com/intern/paste/P2390134443/). The artifact composes `inner_fn` (tensor-subclass wrap/unwrap), `_alias_fn` (output-alias regen), and `_runtime_wrapper` (orchestration; `copy_` reflects the input mutation); `GRAD_PARAM_INDICES = []`.
- **Training step (functionalized RNG + input mutation + output aliasing + backward)**: [generator](https://www.internalfb.com/intern/paste/P2389160349/) -> [generated `python_code`](https://www.internalfb.com/intern/paste/P2390134483/). The artifact composes `_functionalized_rng_wrapper`, `_alias_fn`, and `_runtime_wrapper`, with `GRAD_PARAM_INDICES = [0, 1]` (the driver scatters the harvested grads onto `model.parameters().grad`).

## Example: dynamic (unbacked) shapes

Shapes are static by default; marking a dim with `torch._dynamo.decorators.mark_unbacked` on the input (before calling precompile) captures it as an UNBACKED symint, so ONE artifact runs across runtime sizes of that dim. In the generated `python_code` the symbol renders as `u0` (the `u` prefix means unbacked -- a backed dynamic dim would be `s0`): it is read off the runtime input (`u0 = arg..._size[0]`) and flows through `assert_size_stride(..., (u0, 16), ...)`, the buffer allocations (`empty_strided_cuda((u0, 32), ...)`), and the dynamic `xnumel` that drives the Triton launch (`...xnumel = 32*u0`) -- nothing bakes a constant batch. (The matmul itself runs via `extern_kernels.mm`/`addmm`; the Triton kernel fuses the bias-add + relu.) Because the symbol is unbacked, it cannot be guarded on, which is what makes a no-recompile artifact safe: a graph that needs to guard on a marked dim fails loudly at capture instead of baking a wrong result (see the third example).

All examples are on CUDA; each links the **generator** and the resulting **generated `python_code`** (pastry links for now):

- **Inference, dynamic batch**: [generator](https://www.internalfb.com/intern/paste/P2389226201/) -> [generated `python_code`](https://www.internalfb.com/intern/paste/P2390134523/). An MLP whose batch dim is marked unbacked: `USER_INPUT_SHAPES = [(None, 16)]` and a single `u0` drives the asserts, buffer allocations, and the fused bias-add+relu Triton kernel. The dim is constrained only `u0 >= 0` (plain unbacked, not size-like), so the reloaded artifact runs batch 4 / 32 / 1 / 0, each matching eager.
- **Training, matched batch via `shape_id`**: [generator](https://www.internalfb.com/intern/paste/P2389226227/) -> [generated `python_code`](https://www.internalfb.com/intern/paste/P2390134561/). A loss-vs-target step expects `model(x)` and `target` to share a batch. Marking each unbacked WITHOUT a shape_id gives two independent symbols (`u0`, `u1`) -- the trace succeeds but the batches are treated as unrelated. `mark_unbacked(..., shape_id="b")` instead makes both reuse ONE symbol (`USER_INPUT_SHAPES = [(None, 16), (None, 8)]`, the same `u0` asserted on both inputs), so a mismatched-batch runtime call is rejected. `GRAD_PARAM_INDICES = [0, 1]` harvests the grads onto `model.parameters().grad` for any runtime batch.
- **Safety: a guard-needing dim fails at capture**: [generator](https://www.internalfb.com/intern/paste/P2389226254/) -> [output](https://www.internalfb.com/intern/paste/P2390134594/). A model that branches on the marked dim (`if x.shape[0] > 4`) cannot be captured: an unbacked symint cannot be guarded on, so precompile raises `PrecompileError` at capture (`Could not guard on data-dependent expression u0 > 4`) rather than baking a silently-wrong artifact. This is the no-recompile safety property.

Authored with an AI assistant.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by bobrenjc93** [comment]
Closing: landed on main split across #187859, #188376, #189620, #189621, #189165 and #190226.
